File: exercises/06-capstone/solution/activities.py
import requests
from temporalio import activity
from temporalio.exceptions import ApplicationError

ADNET = "http://localhost:9999"


@activity.defn
def validate_creative(campaign: str) -> str:
    resp = requests.post(f"{ADNET}/validate-creative", json={"campaign": campaign}, timeout=5)
    resp.raise_for_status()
    return resp.json()["creative_id"]


@activity.defn
def reserve_budget(campaign: str) -> str:
    resp = requests.post(f"{ADNET}/reserve-budget", json={"campaign": campaign}, timeout=5)
    resp.raise_for_status()
    return resp.json()["reservation_id"]


@activity.defn
def publish_to_channel(campaign: str, channel: str) -> str:
    activity.logger.info("Publishing to %s (attempt %d)", channel, activity.info().attempt)
    resp = requests.post(f"{ADNET}/publish/{channel}", json={"campaign": campaign}, timeout=5)
    if 400 <= resp.status_code < 500:
        raise ApplicationError(
            f"{channel} rejected the campaign: {resp.json().get('error', resp.text)}",
            type="ChannelPolicyError",
            non_retryable=True,
        )
    resp.raise_for_status()
    return resp.json()["placement_id"]


@activity.defn
def generate_launch_report(campaign: str, creative_id: str, reservation_id: str,
                           placements: list[str]) -> dict:
    return {
        "campaign": campaign,
        "status": "LIVE",
        "creative_id": creative_id,
        "reservation_id": reservation_id,
        "placements": placements,
        "channels_live": len(placements),
    }

File: tracks/typescript/exercises/06-capstone/solution/activities.ts
import { activityInfo } from '@temporalio/activity';
import { ApplicationFailure } from '@temporalio/common';

const ADNET = 'http://localhost:9999';

export interface LaunchReport {
  campaign: string;
  status: 'LIVE';
  creative_id: string;
  reservation_id: string;
  placements: string[];
  channels_live: number;
}

async function postJson(
  path: string,
  body: Record<string, unknown>,
): Promise<{ response: Response; body: Record<string, unknown> }> {
  const response = await fetch(`${ADNET}${path}`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(5_000),
  });
  const parsed = (await response.json()) as Record<string, unknown>;
  return { response, body: parsed };
}

export async function validateCreative(campaign: string): Promise<string> {
  const { response, body } = await postJson('/validate-creative', { campaign });
  if (!response.ok) throw new Error(`${response.status}: ${JSON.stringify(body)}`);
  return body.creative_id as string;
}

export async function reserveBudget(campaign: string): Promise<string> {
  const { response, body } = await postJson('/reserve-budget', { campaign });
  if (!response.ok) throw new Error(`${response.status}: ${JSON.stringify(body)}`);
  return body.reservation_id as string;
}

export async function publishToChannel(
  campaign: string,
  channel: string,
): Promise<string> {
  console.log(`Publishing to ${channel} (attempt ${activityInfo().attempt})`);
  const { response, body } = await postJson(`/publish/${channel}`, { campaign });
  if (response.status >= 400 && response.status < 500) {
    throw ApplicationFailure.nonRetryable(
      `${channel} rejected the campaign: ${body.error ?? response.statusText}`,
      'ChannelPolicyError',
    );
  }
  if (!response.ok) throw new Error(`${response.status}: ${JSON.stringify(body)}`);
  return body.placement_id as string;
}

export async function generateLaunchReport(
  campaign: string,
  creativeId: string,
  reservationId: string,
  placements: string[],
): Promise<LaunchReport> {
  return {
    campaign,
    status: 'LIVE',
    creative_id: creativeId,
    reservation_id: reservationId,
    placements,
    channels_live: placements.length,
  };
}
